File: src/Unit-1/DoThis/Program.cs
using Akka.Actor;

namespace WinTail
{
    #region Program
    class Program
    {
        public static ActorSystem MyActorSystem;

        static void Main()
        {
            // initialize MyActorSystem
            MyActorSystem = ActorSystem.Create("MyActorSystem");

            // Initialize the actors using separate Props.
            var consoleWriterProps = Props.Create<ConsoleWriterActor>(); // **NOT** recommended
            var consoleWriterActor = MyActorSystem.ActorOf(consoleWriterProps, "consoleWriterActor");

            // Make the tail coordinator actor.
            var tailCoordinatorProps = Props.Create(() => new TailCoordinatorActor());
            MyActorSystem.ActorOf(tailCoordinatorProps, "tailCoordinatorActor"); // created via side-effect

            // Pass the tail coordinator actor to fileValidatorActorProps (just adding on extra argument)
            var fileValidatorActorProps = Props.Create(() => new FileValidatorActor(consoleWriterActor));
            MyActorSystem.ActorOf(fileValidatorActorProps, "validationActor"); // created via side-effect

            var consoleReaderProps = Props.Create<ConsoleReaderActor>();
            var consoleReaderActor = MyActorSystem.ActorOf(consoleReaderProps, "consoleReaderActor");
            
            // tell console reader to begin
            consoleReaderActor.Tell(ConsoleReaderActor.StartCommand);

            // blocks the main thread from exiting until the actor system is shut down
            MyActorSystem.WhenTerminated.Wait();
        }
    }
    #endregion
}

File: src/Unit-1/FSharp/DoThis/Program.fs
open System
open Akka.Actor
open Akka.FSharp
open WinTail.Actors
open WinTail.Messages

[<EntryPoint>]
let main _ = 
    let myActorSystem = System.create "MyActorSystem" (Configuration.load())
        
    let consoleWriterActor = spawn myActorSystem "consoleWriterActor" (actorOf consoleWriterActor)
    
    // Supervision strategy used by tailCoordinatorActor
    let strategy () =
        Strategy.OneForOne((fun ex ->
            match ex with
            | :? ArithmeticException -> Directive.Resume
            | :? NotSupportedException -> Directive.Stop
            | _ -> Directive.Restart),
            10, // maximum of 10 restarts
            TimeSpan.FromSeconds(30.)) // within 30 seconds
    
    let tailCoordinatorActor =
        spawnOpt myActorSystem
            "tailCoordinatorActor"
            (actorOf2 tailCoordinatorActor)
            [ SpawnOption.SupervisorStrategy (strategy ()) ]
            
    // Pass tailCoordinatorActor to fileValidatorActor props (adding just one argument)
    let validationActor = spawn myActorSystem "validationActor"
                              (actorOf2 (fileValidationActor consoleWriterActor tailCoordinatorActor))
    let consoleReaderActor =
        spawn myActorSystem "consoleReaderActor" (actorOf2 (consoleReaderActor validationActor))

    consoleReaderActor <! Start

    myActorSystem.WhenTerminated.Wait ()
    0
